### PR TITLE
Better handling of distribution with fractions

### PR DESCRIPTION
**Pull request by lexiross**
Addresses #138 

This is just a first stab - happy to change around anything that should be different!

### CODE REVIEW

File: lib/simplifyExpression/distributeSearch/index.js [hunk 1]
```diff
@@ -197,6 +216,14 @@ function distributeTwoNodes(firstNode, secondNode) {
   return Node.Creator.parenthesis(Node.Creator.operator('+', newArgs));
 }
 
+function hasFraction(args) {
+  return args.filter(isFraction).length > 0;
+}
+
+function isFraction(node) {
```
Comment (lexiross) [inline lib/simplifyExpression/distributeSearch/index.js]: I thought this might be a util function somewhere but I couldn't find it. Do you think maybe we should export functions like this in some kind of util so they can be used around the codebase?

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ const distributeSearch = require('../../../lib/simplifyExpression/distributeSear
 
 const TestUtil = require('../../TestUtil');
 
-function testDistributeMinus(exprStr, outputStr) {
+function testDistributeResult(exprStr, outputStr) {
```
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: I renamed this function because it seemed more general than just testing the `minus` operation - does that seem okay?

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js [hunk 3]
```diff
@@ -50,3 +50,24 @@ describe('distribute', function () {
...
+
+describe('distribute with fractions', function () {
+  const tests = [
+    // distribute the non-fraction term into the numerator(s)
+    ['(1 / x +  x^2) * (x^3 + 1)',
+      '(1 * (x^3 + 1) / x + (x^5 + x^2))',
+    ],
+
+    // same
+    ['(3 / x^2 + x / (x^2 + 3)) * (x^2 + 3)',
```
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: example from #138

**Comment by evykassirer** [comment]
woohoo! thanks

I'm studying for a test today, but hope to review this before end of day tomorrow :)

File: lib/simplifyExpression/distributeSearch/index.js (hunk 1 above)
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: there are some similar functions in `Node.Type`, so you could add it there if you want

File: lib/simplifyExpression/distributeSearch/index.js [hunk 4]
```diff
@@ -197,6 +216,14 @@ function distributeTwoNodes(firstNode, secondNode) {
...
+  return Node.Type.isOperator(node) && node.op === '/';
```
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: you can just use the function `Node.Type.isOperator(node, '/')` - I find that a little cleaner :)

File: lib/simplifyExpression/distributeSearch/index.js [hunk 5]
```diff
@@ -175,9 +175,28 @@ function distributeTwoNodes(firstNode, secondNode) {
   // the new operands under addition, now products of terms
   const newArgs = [];
 
+  // if exactly one group contains at least one fraction, multiply the
```
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: wow I love how clean this change is - just a single block of code 😍

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 2 above)
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: yes! 

We had a big change go through recently that changed a lot of the tests and not all of them were named well- I thought I caught and fixed most of them but I should probably do a pass. 

thanks :)

Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: actually could you just call it `testDistribute`

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js [hunk 6]
```diff
@@ -50,3 +50,24 @@ describe('distribute', function () {
   ];
   tests.forEach(t => testDistributeSteps(t[0], t[1]));
 });
+
+describe('distribute with fractions', function () {
```
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: could you also add a section for testing with substeps? because the code you changed does a specific step that's only part of the overall distribution (right? cause for example in your first test it then goes on to multiply x^2 by x^3 and stuff)

File: lib/simplifyExpression/distributeSearch/index.js [hunk 7]
```diff
@@ -175,9 +175,28 @@ function distributeTwoNodes(firstNode, secondNode) {
...
+  // non-fraction group into the numerators of the fraction group
+  if ([firstArgs, secondArgs].filter(hasFraction).length === 1) {
+    const firstArgsFraction = hasFraction(firstArgs);
```
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: slight preference for naming this `firstArgsHasFraction` (to make it clearer it's a boolean) but I can see that it's kinda long, so it's cool if you'd prefer not

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 2 above)
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: cool, will do

File: lib/simplifyExpression/distributeSearch/index.js (hunk 7 above)
Comment (lexiross) [inline lib/simplifyExpression/distributeSearch/index.js]: no I agree, I'll change it!

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 6 above)
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: sure, will do

**Comment by lexiross** [comment]
thanks for the review @evykassirer! and thanks lol - the fat cat thing was quite fun and I had no idea so many people on the internet would be into it 😂 

just updated - lmk if there's anything else I can do

File: lib/simplifyExpression/distributeSearch/index.js [hunk 8]
```diff
@@ -221,7 +221,7 @@ function hasFraction(args) {
 }
 
 function isFraction(node) {
```
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: do you want to move this to `Node.Type`? I could go either way

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js [hunk 9]
```diff
@@ -69,5 +64,21 @@ describe('distribute with fractions', function () {
     ],
   ];
 
-  tests.forEach(t => testDistributeResult(t[0], t[1]));
+  const multiStepTests = [
+
+    ['(2 / x +  3x^2) * (x^3 + 1)',
+      ['(2 * (x^3 + 1) / x + 3x^2 * (x^3 + 1))',
```
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: I'm kinda surprised that there's not parenthesis around `2 * (x^3 + 1)` - what do you think?

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js [hunk 10]
```diff
@@ -69,5 +64,21 @@ describe('distribute with fractions', function () {
...
+        '(2 * (x^3 + 1) / x + (3x^5 + 3x^2))']
+    ],
+
+    ['(2x + x^2) * (1 / (x^2 -4) + 4x^2)',
```
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: could you make one of them have a fancy numerator that need to get multiplied?

e.g. `(2x + x^2) * (3x^2 / (x^2 -4) + 4x^2)`

I'm not sure if this would end up getting multiplied actually... you can see how `(1 * (2x + x^2)` isn't resolved. Do you think it would be pedagogically useful to multiply it out as a substep? (vs how much extra code would it be :p )

File: lib/simplifyExpression/distributeSearch/index.js (hunk 8 above)
Comment (lexiross) [inline lib/simplifyExpression/distributeSearch/index.js]: Oh yeah - I was thinking no since the functions in that class check for more specific types of fractions. I think it would make sense to move this one once it gets defined more than once :)

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 9 above)
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: Oh, good point - I had thought that given `a * b / c`, the `a * b` is considered the numerator. Is that not the case? If not, how would I fix that? Is there a way to explicitly parenthesize an expression? I think on lines 19 and 20 the order of ops is already correct...?

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 10 above)
Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: Okay, I added that as a test (which does end up getting multiplied out). I think the existing test not getting fully distributed is due to the `1` being a constant rather than say `2x`? If we wanted a quick fix for that test, we could just add a rule that `1 * (any expr) = (any expr)`. What do you think?

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 9 above)
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: yeah, that's what I was thinking too. You can explicitly parenthesize it by wrapping it in a `Node.Creator.Parenthesis` node, I hope that fixes this

File: test/simplifyExpression/distributeSearch/distributeSearch.test.js (hunk 10 above)
Comment (evykassirer) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: ah okay, cool

there is a rule for that, but it's not used for distribution simplifications. maybe I'll just open another issue to add that in, since it's not really related to this PR - sound good?

Comment (lexiross) [inline test/simplifyExpression/distributeSearch/distributeSearch.test.js]: sounds good!

**Comment by lexiross** [comment]
@evykassirer added the parenthesis step - but then ended up with an edge case for `2(x+3)/3`, which should have solved to `2/3 x + 2` but was instead `2x / 3 + 2`. I decided to only do the parentheses check if the denominator of the fraction we're working with is NOT a constant. Does that sound right?

File: lib/simplifyExpression/distributeSearch/index.js [hunk 11]
```diff
@@ -178,13 +178,19 @@ function distributeTwoNodes(firstNode, secondNode) {
   // if exactly one group contains at least one fraction, multiply the
   // non-fraction group into the numerators of the fraction group
   if ([firstArgs, secondArgs].filter(hasFraction).length === 1) {
+    // console.log("FIRST ARGS", firstArgs);
```
Comment (evykassirer) [inline lib/simplifyExpression/distributeSearch/index.js]: (remove these before we merge)

**Comment by evykassirer** [comment]
Hmmm yeah we should maybe standardize if we do `2/3 x` or `2x / 3` but I believe both can show up in the solver and I don't really mind if it ends up one way or the other. I'll open an issue to discuss that!

So I think it's alright either way you do it (though I have slight preference for always adding the parens)

Thanks for your patience this week btw :) It's been really heavy workload for me so I haven't been able to be on GitHub as much, but you're bringing up lots of good points and also this diff is great! :D

File: lib/simplifyExpression/distributeSearch/index.js (hunk 11 above)
Comment (lexiross) [inline lib/simplifyExpression/distributeSearch/index.js]: whoops, my bad 🙈

**Comment by lexiross** [comment]
Okay! So if we do the parentheses all the time, it's okay to change the existing test?

And no worries at all!! I'm in no rush :)

**Comment by evykassirer** [comment]
Yeah, I don't think it changes the correctness or pedagogy, so go for it :)

**Comment by lexiross** [comment]
Okay, updated :)
